Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,211 @@
+        private readonly IProductSearchService _productSearchService;
+        private readonly ICategorySearchService _categorySearchService;
+        private readonly IPropertySearchService _propertySearchService;
+        private readonly IProperyDictionaryItemSearchService _propertyDictionarySearchService;
+
+        public CatalogFullExportPagedDataSource(ICatalogService catalogService, ICatalogSearchService catalogSearchService, IProductSearchService productSearchService, ICategorySearchService categorySearchService, ICategoryService categoryService,
+                                  IItemService itemService, IPropertyService propertyService, IPropertySearchService propertySearchService, IProperyDictionaryItemSearchService propertyDictionarySearchService,
+                                  IProperyDictionaryItemService propertyDictionaryService, IBlobStorageProvider blobStorageProvider, IAssociationService associationService)
+        {
+            _productSearchService = productSearchService;
+            _categorySearchService = categorySearchService;
+            _propertySearchService = propertySearchService;
+            _propertyDictionarySearchService = propertyDictionarySearchService;
+            _catalogSearchService = catalogSearchService;
+        }
+
+        public int TotalCount { get; set; }
+        public int CurrentPageNumber { get; protected set; }
+        public int PageSize { get; set; } = 50;
+
+        public IEnumerable<IExportable> Items { get; protected set; }
+
+        private bool _statesInitialized = false;
+
+        private readonly ExportDataSourceState[] _exportDataSourceStates = new ExportDataSourceState[5];
+
+        private ExportDataQuery _dataQuery;
+
+        public ExportDataQuery DataQuery
+        {
+            set
+            {
+                _dataQuery = value;
+                CurrentPageNumber = 0;
+                TotalCount = -1;
+                InitExportDataSourceStates();
+                CalculateCounts();
+            }
+        }
+
+        public int GetTotalCount()
+        {
+            CalculateCounts();
+            return TotalCount;
+        }
+
+        private void EnsureHaveTotals()
+        {
+            if (TotalCount < 0)
+            {
+                CalculateCounts();
+            }
+        }
+
+        public bool Fetch()
+        {
+            EnsureHaveTotals();
+
+            var take = _dataQuery.Take ?? PageSize;
+            var skip = _dataQuery.Skip ?? CurrentPageNumber * PageSize;
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
+
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.FetchResult.Result = Array.Empty<IExportable>();
+                var overlap = skip - curStateStartPos; // Positive overlap should fall into skip, negative into take
```

Comment:
This logic seems harder to me than when we have TotalCounts already filled like here:
https://github.com/VirtoCommerce/vc-platform-core/pull/521/files#diff-025b92c5e3684f6c6184640d95e9ba6dR89

Could we replace it (with breaking the cycle when portion is fully filled)?

Reply:
I don't understand. There TotalCounts filled too. Need to discuss.

Reply:
Yes, lets discuss on this.